ASoC: SOF: ipc4-topology/Intel: hda-pcm: Fix ChainDMA host buffer time and move constraint to period time

ujfalusi opened the pull request:
Hi,

ChainDMA PCM use 5ms host buffer by default and not 2ms, this needs to be handled correctly to be able to place correct constraint in platform code.

We should also move the constraint from buffer time to period time - which will implicitly applies the same constraint on the buffer size to make sure that applications will not use smaller periods than the initial DMA burst.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Account for different ChainDMA host buffer size

For ChainDMA the firmware allocates 5ms host buffer instead of the standard
4ms which should be taken into account when setting the constraint on the
buffer size.

Fixes: 72af064faead ("ASoC: SOF: ipc4-topology: Save the DMA maximum burst size for PCMs")
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda-pcm: Place the constraint on period time instead of buffer time

Instead of constraining the ALSA buffer time to be double of the firmware
host buffer size, it is better to set it for the period time.
This will implicitly constrain the buffer time to a safe value
(num_periods is at least 2) and prohibits applications to set smaller
period size than what will be covered by the initial DMA burst.

Fixes: 02ea2a0364a2 ("ASoC: SOF: Intel: hda-pcm: Use dsp_max_burst_size_in_ms to place constraint")
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- comment fixed in patch 2 ("To make avoid immediate xrun ..." -> "To avoid immediate xrun ...")

kv2019i commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -671,8 +671,13 @@ static int sof_ipc4_widget_setup_pcm(struct snd_sof_widget *swidget)
+			struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+			sps->dsp_max_burst_size_in_ms = pipeline->use_chain_dma ?
+				SOF_IPC4_CHAIN_DMA_BUFFER_SIZE : SOF_IPC4_MIN_DMA_BUFFER_SIZE;
```
Not related to this PR, but for all current IPC4 platforms, max burst size is 4ms:
https://github.com/thesofproject/sof/blob/main/zephyr/lib/dma.c#L21

We can't currently query this value (would need to be on a per PCM basis). But if we start hardcoding the actual size for chain-dma in driver (now kernel+FW need to stay aligned), why not do the same for other types of PCMs as well?

ujfalusi replied:
Hrm, the host buffer size is configured by IPC and it is 2ms by default:

https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-topology.h#L74
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-topology.c#L2394

and in FW:
https://github.com/thesofproject/sof/blob/main/src/audio/host-zephyr.c#L934

But are you saying that the 2ms which we set is ignored by fw it is using 4ms instead out of fun? Since when this is done? This is completely new to me...

ujfalusi replied:
The ChainDMA is documented to use this 'calculation' in firmware to get to 5ms from 2ms buffer size but I did missed to capture this in this place.

kv2019i reviewed (dismissed):
Looks good and will improve compatibility with ALSA apps. Added a note on possible way to go even further with this line of thought...

ujfalusi committed:
ASoC: SOF: ipc4-topology: Correct the minimum host DMA buffer size

The firmware has changed the minimum host buffer size from 2 periods to
4 periods (1 period is 1ms) which was missed by the kernel side.

Adjust the SOF_IPC4_MIN_DMA_BUFFER_SIZE to 4 ms to align with firmware.

Link: https://github.com/thesofproject/sof/commit/f0a14a3f410735db18a79eb7a5f40dc49fdee7a7
Fixes: a2db53360203 ("ASoC: SOF: ipc4-topology: Do not parse the DMA_BUFFER_SIZE token")
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v2:
- add a new patch to correct the default host buffer size to 4 ms (thank @kv2019i for noticing it)

ranj063 commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -70,8 +70,8 @@
-/* FW requires minimum 2ms DMA buffer size */
-#define SOF_IPC4_MIN_DMA_BUFFER_SIZE	2
+/* FW requires minimum 4ms DMA buffer size */
+#define SOF_IPC4_MIN_DMA_BUFFER_SIZE	4
```
@ujfalusi does this fix any issues today? In the firmware we always set the dma_buffer_size to be the max of period_size * 4 or the buffer size coming from the kernel. for the host: https://github.com/thesofproject/sof/blob/f0a14a3f410735db18a79eb7a5f40dc49fdee7a7/src/audio/host-zephyr.c#L864

So, just wondering what this change fixes and while I agree we should align with the firmware, it might be redundant

kv2019i replied:
@ranj063  @ujfalusi  I think we have a real problem in setting the minimum period-size correctly. The third patch in v2 is not quite correct as the 4ms magic setting is not universal in FW code, but a platform-specific variable. But in principle, it is a major problem if the host doesn't know the effective buffer size set. Now it knows 2ms is the minimum, but it currently doesn't know wheter 2ms or 4ms was set. What happens now with Pipewire is that we set min-period-size based on 2ms assumption but then when FW in practise does 4ms DMA bursts, this leads to xrun loops (the problem Péter has been trying to address with Pipewire headroom patches). Long term, we'd like to extend ALSA to be able to relay this "max-dma-burst-size" to user-space applications, but to start, we should set min-period-size based on this information. Currently we are not doing this correctly.

ranj063 replied:
Thanks for the context @kv2019i this makes sense now

lgirdwood reviewed (dismissed):
This will fix for now - we can find better HW specific solution for future.

kv2019i commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -70,8 +70,8 @@
 #define SOF_IPC4_CHAIN_DMA_NODE_ID	0x7fffffff
```
Minor nit in commit message: the firmware change was for a particular Intel platforms. But yeah, in practise it covers all platforms covered by this ipc4 kernel code at the moment.

ujfalusi replied:
the fw change was for _all_ IPC4 platforms (_all_ Intel platforms which can run IPC4), it is fair to say that for IPC4 the host buffer size has been 4ms and not 2ms.

kv2019i replied:
Right, that is true, but it's just hard to verify this looking at the commit and then looking at the FW commit. Reader has to have outside information that Intel ACE variant happens to be the only IPC4 supporting platform and the various other platforms in SOF (with 2ms default) are never used in IPC4 mode.

ujfalusi replied:
CAVS2.5 works (but not officially supported) with IPC4 and that also uses 4ms host buffer. At this time IPC4 as such uses 4ms host buffer on all supported platforms.

Yes, we will have support for this information to be queried from firmware and can get rid of this define, but as far as IPC4 concerns, we have been mistakenly assumed 2ms host buffer when in fact IPC4 is using 4ms.

kv2019i commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -291,19 +291,25 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 	 * On playback start the DMA will transfer dsp_max_burst_size_in_ms
```
commit: "double the firmware host DMA buffer size". Otherwise I read this referring to kernel side buffers and doubling them.

Also typo: s/for the priod time/for the period time/

ujfalusi replied:
OK, reworded the commit message.

kv2019i commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -291,19 +291,25 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
+	if (spcm->stream[direction].dsp_max_burst_size_in_ms) {
+		unsigned int period_time = spcm->stream[direction].dsp_max_burst_size_in_ms;
+
+		period_time += min(period_time, 10);
```
Should this 10 be a define?

ujfalusi replied:
I have hard time to come up with a define name, should a comment be sufficient?
```c
		/*
		 * add maximum of 10ms headroom over the maximum burst size to
		 * cover the time needed for the DMA pace to settle
		 */
```

kv2019i commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -291,19 +291,25 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 		snd_pcm_hw_constraint_minmax(substream->runtime,
-			SNDRV_PCM_HW_PARAM_BUFFER_TIME,
-			spcm->stream[direction].dsp_max_burst_size_in_ms * USEC_PER_MSEC * 2,
+			SNDRV_PCM_HW_PARAM_PERIOD_TIME,
```
Did you test more extreme deep-buffer cases? I think this should work, e.g. 50ms deep buffer would result in 60ms minimum period size, which should be ok.

ujfalusi replied:
I did run some test with aplay, mpv and they worked as before.

bardliao replied:
According to
```c
#define	SNDRV_PCM_HW_PARAM_PERIOD_TIME	12	/* Approx distance between
						 * interrupts in us
```
SNDRV_PCM_HW_PARAM_PERIOD_TIME is in `us`, but `period_time` here seems to be in `ms`. Does it correct?

ujfalusi replied:
It is converted to usec with `period_time * USEC_PER_MSEC`

kv2019i reviewed (commented):
Patches look great, I think this is the right approach. Some minor nits as inline comments.

ujfalusi commented:
Changes since v4:
- use define for the 10ms period time headroom
- reword the commit message of patch 3
